### DISCUSSION TITLE
Write value formatting

### DISCUSSION THREAD

**Discussion by donnie-kerr** [capacitor-community/bluetooth-le#449] [General]
Newbie question.  A chinese mfg told me to write a command from my app to the watch in these exact words:
The APP write 0xD0 command to the watch via the UUID F0080003, and the parameter configuration = 0x01. 
Full command: 0xD001
service:F0080001-0451-4000-B000-000000000000
write characteristic:F0080003-0451-4000-B000-000000000000
value:0xD001
My question is how do I format this value as a DataView to write it using the write function?
` await BleClient.write(device.deviceId, srv, char, value??)`

**Comment by donnie-kerr** [comment]
Saw an example like this.  Is this correct way to do it? 

`let hex = 0xD001;
 let octets = Math.floor(hex.replace(/^0x/, '').length / 2)
let dataView = new DataView(new ArrayBuffer(octets))`

**Comment by peitschie** [comment]
You can construct numbers directly from hex values in JavaScript. In theory, what you need is as simple as:
```
await BleClient.write(device.deviceId, srv, char, numbersToDataView([0xD0, 0x01]))
```

(using the `numbersToDataView` utility method mentioned at https://github.com/capacitor-community/bluetooth-le#write)

**Reply by donnie-kerr** [reply]
Cool. Thank you
Related question. Is It is possible to go the other way from a dataview value to hex? 

When a notification comes in the value looks like this:
a1 00 00 00 23 2b 00 05 08 00 00 01 00 00 00 00 00 00 00 00
they tell me it should be this hex:
0xd04f000000000000000000000000000000000000 
where the 4f is the heart rate measurement I want. 
So I'm wondering if this value returned by startnotification is DataView and needs to be converted to hex? Is it a byteArray that needs converted to hex?

**Reply by peitschie** [reply]
Just a quick google, but something like this should work for you: https://stackoverflow.com/a/40031979/156169

```
function buf2hex(buffer) { // buffer is an ArrayBuffer
  return [...new Uint8Array(buffer)]
      .map(x => x.toString(16).padStart(2, '0'))
      .join('');
}

// EXAMPLE:
const buffer = new Uint8Array([ 4, 8, 12, 16 ]).buffer;
console.log(buf2hex(buffer)); // = 04080c10
```

You can get the buffer from a `DataView` via the `.buffer` property: https://devdocs.io/javascript/global_objects/dataview/buffer

**Reply by donnie-kerr** [reply]
here is my log in xcode where I see the value = a1 00 00 00 23 2b 00 05 08 00 00 01 00 00 00 00 00 00 00 00 .  Is this the buffer inside the returned DataView? 
<img width="895" alt="Screenshot 2022-12-01 at 5 06 36 PM" src="https://user-images.githubusercontent.com/116496781/205200259-02cf848b-36fb-45da-b79e-def31ff67515.png">

**Reply by peitschie** [reply]
I don't have any great insights there @rpm-one , sorry.

**Comment by donnie-kerr** [comment]
buf2hex looks like it just took the dataview.buffer and removed the spaces.
<img width="890" alt="Screenshot 2022-12-02 at 7 03 09 AM" src="https://user-images.githubusercontent.com/116496781/205299123-bda1a9b7-5c86-42e1-b480-cbf5bf9b2035.png">

**Comment by donnie-kerr** [comment]
Well, 
Full command: 0xD001 write seems to be successful and startNotifications seems to start.  But I do not receive anything from the device. Don't get enough information to know if the write actually worked or if command is correct.  It is supposed to start the device sending, but it doesn't. No idea why.  Waiting on mfg to help.  But, if you see anything below that may be an issue, please let me know. I don't know what the JS undefined messages mean. 

`let srv = 'f0080001-0451-4000-b000-000000000000'
let char = 'f0080003-0451-4000-b000-000000000000'
await BleClient.write(device.deviceId, srv, char, numbersToDataView([0xD0, 0x01]))`

`resolve connect Connection successful.
⚡️  TO JS undefined
⚡️  [log] - connected to device {"name":"P8","deviceId":"BB84AA72-8984-61DA-58BB-C912943525D7"}
⚡️  To Native ->  BluetoothLe write 68670559
Resolve write|F0080001-0451-4000-B000-000000000000|F0080003-0451-4000-B000-000000000000 Successfully written value.
⚡️  TO JS undefined
⚡️  To Native ->  BluetoothLe addListener 68670560
⚡️  To Native ->  BluetoothLe startNotifications 68670561
Set notifications true
Resolve setNotifications|F0080001-0451-4000-B000-000000000000|F0080002-0451-4000-B000-000000000000 Successfully set notifications.
⚡️  TO JS undefined`
